tests/device_compiler/parallel_for_class_ctor_and_dtor.cpp

```diff
@@ -0,0 +1,87 @@
+#include <CL/sycl.hpp>
+#include <iostream>
+#include <numeric>
+#include <vector>
+
+#include <boost/test/minimal.hpp>
+
+/*
+  A contrived test with the aim of testing that C++ objects can be created and
+  destroyed inside a kernel implicitly and to detect any problems with
+  address spaces that may occur in the process.
+
+  Current State:
+    Works with POCL (tested with top of tree (10/13/2018) triSYCL device
+    compiler)
+    Untested with XOCL
+*/
+
+using namespace cl::sycl;
+
+constexpr size_t N = 32;
+
+class Index_Data {
+public:
+  Index_Data() : g(0), r(0) {};
+  ~Index_Data() {};
+
+  std::size_t g;
+  std::size_t r;
+
+};
+
+class Index_Member_Data {
+  std::size_t g;
+  std::size_t r;
+
+public:
+
+  Index_Member_Data() : g(0), r(0) {};
+  ~Index_Member_Data() {};
+
+  auto get_g() { return g; }
+  auto get_r() { return r; }
+  void set_g(std::size_t g_) { g = g_; }
+  void set_r(std::size_t r_) { r = r_; }
+};
+
+int test_main(int argc, char *argv[]) {
+  queue my_queue{default_selector{}};
+
+  buffer<unsigned int, 1> a{N};
+
+  auto acc_w = a.get_access<access::mode::write>();
+  for (unsigned int i = 0; i < N; ++i) {
+    acc_w[i] = 0;
+  }
+
+  Index_Member_Data idm_outer;
+  idm_outer.set_g(20);
+  idm_outer.set_r(20);
+
+  my_queue.submit([&](handler &cgh) {
+    auto acc = a.get_access<access::mode::write>(cgh);
+
+    Index_Member_Data idm_inner;
+    idm_inner.set_g(10);
+    idm_inner.set_r(10);
+
+    cgh.parallel_for<class ctor_dtor>(
+        range<1>{N}, [&, d_a = drt::accessor<decltype(acc)>{acc}](item<1> index) {
```

Comment:
I do not think you can use `&` on a device where there is no unified memory with the host.
Furthermore, even on CPU, the kernel is executed in another thread and `idm_inner` is probably destructed already when the kernel start.
This is why in SYCL we have only capture by value.